packages/typespec-ts/test/modularUnit/scenarios/multipart/file.md

```diff
@@ -0,0 +1,187 @@
+
+```ts operations
+import { TestingContext as Client } from "./index.js";
+import { RequestBody, requestBodySerializer } from "../models/models.js";
+import {
+  StreamableMethod,
+  PathUncheckedResponse,
+  createRestError,
+  operationOptionsToRequestParameters,
+} from "@azure-rest/core-client";
+
+export function _doThingSend(
+  context: Client,
+  bodyParam: RequestBody,
+  options: DoThingOptionalParams = { requestOptions: {} },
+): StreamableMethod {
+  return context
+    .path("/")
+    .post({
+      ...operationOptionsToRequestParameters(options),
+      contentType: "multipart/form-data",
+      body: requestBodySerializer(bodyParam),
+    });
+}
+
+export async function _doThingDeserialize(
+  result: PathUncheckedResponse,
+): Promise<void> {
+  const expectedStatuses = ["204"];
+  if (!expectedStatuses.includes(result.status)) {
+    throw createRestError(result);
+  }
+
+  return;
+}
+
+export async function doThing(
+  context: Client,
+  bodyParam: RequestBody,
+  options: DoThingOptionalParams = { requestOptions: {} },
+): Promise<void> {
+  const result = await _doThingSend(context, bodyParam, options);
+  return _doThingDeserialize(result);
+}
+```
+
+# File part, filename required
+
+```tsp
+model FileRequiredName extends File {
+  filename: string;
+}
+
+model RequestBody {
+  nameRequired: HttpPart<FileRequiredName>;
+}
+
+op doThing(@header contentType: "multipart/form-data", @multipartBody bodyParam: RequestBody): void;
+```
+
+## Models
+
+The filename must be provided _somehow_. This can either be done by passing a `File` object, which has a required filename property, or by using the wrapper object to pass a `filename` alongside the `contents`.
+
+```ts models
+import {
+  FileContents,
+  createFilePartDescriptor,
+} from "../static-helpers/multipartHelpers.js";
+
+/** model interface RequestBody */
+export interface RequestBody {
+  nameRequired:
+    | File
+    | { contents: FileContents; contentType?: string; filename: string };
+}
+
+export function requestBodySerializer(item: RequestBody): any {
+  return [
+    createFilePartDescriptor("nameRequired", item["nameRequired"], undefined),
+  ];
+}
+```
+
+# only: Default content type
+
+```tsp
+model PngFile extends File {
+  contentType: "image/png";
+}
+
+model RequestBody {
+  image: HttpPart<PngFile>;
+}
+
+op doThing(@header contentType: "multipart/form-data", @multipartBody bodyParam: RequestBody): void;
+```
+
+## Models
+
+```ts models
+import {
+  FileContents,
+  createFilePartDescriptor,
+} from "../static-helpers/multipartHelpers.js";
+
+/** model interface RequestBody */
+export interface RequestBody {
+  image:
+    | FileContents
+    | { contents: FileContents; contentType?: "image/png"; filename?: string };
+}
+
+export function requestBodySerializer(item: RequestBody): any {
+  return [createFilePartDescriptor("image", item["image"], "image/png")];
+}
+```
+
+# only: Multiple files
+
+```tsp
+model RequestBody {
+  files: HttpPart<File>[];
```

Comment:
should we add a case for 
```
model RequestBody {
  files: HttpPart<File[]>;
}
```

Reply:
I don't think that case really makes sense. That body model would represent a single part consisting of a JSON array of `File` objects.

We do already have a case for a JSON array [here](https://github.com/Azure/autorest.typescript/pull/2984/files#diff-c2c7ad81c14f6edffaa7c0de7676cb537f932a6eb4e75d8ebc03ffbf738c0eb2R48-R65).